File: backend/services/stigmer-server/pkg/domain/workflowexecution/temporal/dispatch.go
package temporal

import (
	"github.com/rs/zerolog/log"
	sessionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/session/v1"
)

// WfExecQueuePrefix is the prefix for per-execution task queue names.
// Format: "wfexec:{workflow_execution_id}"
const WfExecQueuePrefix = "wfexec:"

// WorkflowDispatchResult holds the resolved Temporal task queue and execution
// target for the workflow execution's child workflow dispatch.
type WorkflowDispatchResult struct {
	TaskQueue       string
	ExecutionTarget sessionv1.ExecutionTarget
}

// FormatWfExecTaskQueue derives the canonical Temporal task queue name for a
// given workflow execution ID. Format: "wfexec:{execution_id}".
func FormatWfExecTaskQueue(executionID string) string {
	return WfExecQueuePrefix + executionID
}

// ResolveWorkflowTaskQueue determines which Temporal task queue a workflow
// execution's child workflow should be started on.
//
// Routing modes (controlled by Config.WorkflowActivityRouting):
//   - "global": Always returns Config.RunnerQueue (stigmer_runner). All workflow
//     executions share the global runner pool. Default for OSS.
//   - "execution": Returns wfexec:{execution_id} when execution_target resolves
//     to CLOUD. A dedicated sandbox is provisioned for that queue. Used by
//     managed cloud service for sandbox sharing between workflow and child agents.
func ResolveWorkflowTaskQueue(executionID string, executionTarget sessionv1.ExecutionTarget, cfg *Config) WorkflowDispatchResult {
	resolved := resolveWorkflowExecutionTarget(executionTarget, cfg)

	if cfg.WorkflowActivityRouting == RoutingExecution && resolved == sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD {
		taskQueue := FormatWfExecTaskQueue(executionID)

		log.Info().
			Str("execution_id", executionID).
			Str("task_queue", taskQueue).
			Str("routing_mode", cfg.WorkflowActivityRouting).
			Str("execution_target", resolved.String()).
			Msg("Workflow dispatch resolved per-execution queue")

		return WorkflowDispatchResult{
			TaskQueue:       taskQueue,
			ExecutionTarget: resolved,
		}
	}

	log.Info().
		Str("execution_id", executionID).
		Str("task_queue", cfg.RunnerQueue).
		Str("routing_mode", cfg.WorkflowActivityRouting).
		Str("execution_target", resolved.String()).
		Msg("Workflow dispatch resolved global queue")

	return WorkflowDispatchResult{
		TaskQueue:       cfg.RunnerQueue,
		ExecutionTarget: resolved,
	}
}

// resolveWorkflowExecutionTarget resolves UNSPECIFIED to the configured default.
func resolveWorkflowExecutionTarget(target sessionv1.ExecutionTarget, cfg *Config) sessionv1.ExecutionTarget {
	if target != sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED {
		return target
	}
	if cfg.DefaultExecutionTarget == DefaultExecutionTargetCloud {
		return sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD
	}
	return sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL
}

File: backend/services/stigmer-server/src/temporal/workflowexecution/dispatch.ts
/**
 * Workflow dispatch-queue resolution — ports
 * pkg/domain/workflowexecution/temporal/dispatch.go
 * (ResolveWorkflowTaskQueue): which Temporal task queue a workflow
 * execution's CHILD workflow starts on.
 *
 * Unlike agentexecution's dispatch (which loads the session to read its
 * harness and target), this resolution is PURE — the execution target
 * rides in on the create pipeline's spec read (the seam's
 * StartWorkflowExecutionInput.executionTarget), so there is no store
 * read and no failure lane: Go's Create has no dispatch error boundary
 * and neither does this port.
 *
 * Routing modes (config.workflowActivityRouting):
 *   - "global": always the configured runner queue (stigmer_runner) —
 *     all workflow executions share the global runner pool.
 *   - "execution": wfexec:{execution_id} regardless of execution target.
 *     For CLOUD a dedicated sandbox is provisioned for that queue; for
 *     LOCAL the desktop runner-manager creates a worker on it.
 */
import type { ExecutionTarget } from "@stigmer/protos/ai/stigmer/agentic/session/v1/enum_pb";

import type { Logger } from "../../boot/logger.js";
import {
  WORKFLOW_ROUTING_EXECUTION,
  type WorkflowExecutionTemporalConfig,
} from "../../domain/workflowexecution/temporal/config.js";
import { formatWfExecTaskQueue } from "./names.js";

/**
 * The resolved Temporal task queue and effective execution target for a
 * workflow execution's child dispatch (Go WorkflowDispatchResult).
 */
export interface WorkflowDispatchResult {
  readonly taskQueue: string;
  readonly executionTarget: ExecutionTarget;
}

/**
 * Go ResolveWorkflowTaskQueue: resolves the child's task queue from the
 * routing mode, and UNSPECIFIED targets through the config's single
 * resolution rule.
 */
export function resolveWorkflowTaskQueue(
  executionId: string,
  executionTarget: ExecutionTarget,
  config: WorkflowExecutionTemporalConfig,
  logger: Logger,
): WorkflowDispatchResult {
  const resolved = config.resolveWorkflowExecutionTarget(executionTarget);

  if (config.workflowActivityRouting === WORKFLOW_ROUTING_EXECUTION) {
    const taskQueue = formatWfExecTaskQueue(executionId);
    logger.info("Workflow dispatch resolved per-execution queue", {
      execution_id: executionId,
      task_queue: taskQueue,
      routing_mode: config.workflowActivityRouting,
      execution_target: resolved,
    });
    return { taskQueue, executionTarget: resolved };
  }

  logger.info("Workflow dispatch resolved global queue", {
    execution_id: executionId,
    task_queue: config.runnerQueue,
    routing_mode: config.workflowActivityRouting,
    execution_target: resolved,
  });
  return { taskQueue: config.runnerQueue, executionTarget: resolved };
}
